qspectra/dynamics/heom.py

```diff
@@ -0,0 +1,308 @@
+        bath_coeffs.append(4 * reorg_en * gamma * matsu_freqs[m] / (inv_T *
+                         (matsu_freqs[m] ** 2 - gamma ** 2)))
+    return bath_coeffs
+
+def HEOM_tensor(hamiltonian, subspace='ge', M=3, level_cutoff=3):
+    """
+    Calculates the HEOM tensor elements in the energy eigenbasis
+
+    the Liouville space is: (H^k) \tensordot (H^k)
+
+    where H is the Hilbert space of the given subspace, and k is the total
+    number of auxilary density matrices + 1 (the original density matrix)
+
+    If the hilbert space is n x n, the total shape of the tensor is:
+
+    [n*n*m,n*n*m]
+
+
+        for n = 2, level_cutoff = 3
+
+        p_00  p_01  p_10  p_11
+    p_00[   ]|[   ]|[   ]|[   ]
+        [   ]|[   ]|[   ]|[   ]
+        -----|-----|-----|-----
+    p_01[   ]|[   ]|[   ]|[   ]
+        [   ]|[   ]|[   ]|[   ]
+        -----|-----|-----|-----
+    p_10[   ]|[   ]|[   ]|[   ]
+        [   ]|[   ]|[   ]|[   ]
+        -----|-----|-----|-----
+    p_11[   ]|[   ]|[   ]|[   ]
+        [   ]|[   ]|[   ]|[   ]
+
+    Parameters
+    ----------
+    hamiltonian : hamiltonian.Hamiltonian
+        Hamiltonian object specifying the system
+    subspace : container, default 'ge'
+        Container of any or all of 'g', 'e' and 'f' indicating the desired
+        subspaces on which to calculate the HEOM tensor
+    M : cutoff for the number of exponential functions to include in the
+        correlation function
+        (note: summation is 0-indexed, so M+1 exponentials are included)
+    level_cutoff : level at which heiarchy is truncated
+
+    Returns
+    -------
+    out : np.ndarray
+        tensor giving the HEOM transfer rates between the
+        density matrix elements and aux density matrix elements
+        in the system energy eigenbasis
+
+    References
+    # 10.1021/jp109559p (indexing nomenclature refers to eqn. 9)
+    doi: 10.1063/1.3271348 (eqn. A22)
+    ----------
+    """
+
+    N = hamiltonian.n_states(subspace)
+    gamma = hamiltonian.bath.cutoff_freq
+    temp = hamiltonian.bath.temperature
+    reorg_en = hamiltonian.bath.reorg_energy
+
+    matsu_freqs = matsubara_frequencies(M, gamma, temp)
+    bath_coeffs = corr_func_coeffs(M, gamma, temp, reorg_en, matsu_freqs)
+
+    rho_indices, mat_to_ind = ADO_mappings(N, M, level_cutoff)
+    tot_rho = len(rho_indices)
+    print 'there are {} ADOs in total'.format(tot_rho)
+
+    L = lil_matrix((tot_rho * N * N, tot_rho * N * N), dtype=np.complex128)
+
+    # Tensor variables:
+    Isq = np.eye(N * N)
+    Nsq = N ** 2
+
+    # unitary evolution:
+    H = np.diag(hamiltonian.E(subspace))
+    liouvillian = -1j * super_commutator_matrix(H)
+
+    # precompute the list of N vectorized projection operators
+    proj_op_left = []
+    proj_op_right = []
+    print 'creating projection operators'
+    for n in xrange(N):
+        proj_op = np.zeros((N, N))
+        proj_op[n, n] = 1
+        proj_op_left.append(super_left_matrix(proj_op))
```

Comment:
fixed in https://github.com/shoyer/qspectra/commit/6acac1ecc50f2260c068c01906c4354df227476c,
i haven't profiled this yet, it's possible that the matrices that we use to build the tensor are not that sparse to begin with